entry_types/scrolled/package/spec/frontend/MediaPlayer-spec.js

```diff
@@ -30,41 +32,103 @@ describe('MediaPlayer', () => {
+    let state = getInitialPlayerState();
+    let playerActions = getPlayerActions();
+    const {rerender} = render(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+    
+    state = getInitialPlayerState();
+    state.shouldPlay = true;
+
+    let spyPlay = jest.spyOn(media.players['0'], 'play');
+    rerender(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+    expect(spyPlay).toHaveBeenCalledTimes(1);
   });
 
+  // postpone till volumebinding is added to videoPlayer
+  // it('calls playAndFadeIn on player when shouldPlay changes to true and fadeDuration is present', () => {
+  //   let state = getInitialPlayerState();
+  //   let playerActions = getPlayerActions();
+  //   const {rerender} = render(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+    
+  //   state = getInitialPlayerState();
+  //   state.fadeDuration = 100;
+  //   state.shouldPlay = true;
+  //   console.log(media.players['0'].playAndFadeIn);
+    
+  //   let spyPlayAndFadeIn = jest.spyOn(media.players['0'], 'playAndFadeIn');
+  //   rerender(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+  //   expect(spyPlayAndFadeIn).toHaveBeenCalledTimes(1);
+  //   expect(spyPlayAndFadeIn).toHaveBeenCalledWith(100);
+  // });
+
+  it('calls pause on player when isPlaying and shouldPlay changes to false in playerState', () => {
+    let state = getInitialPlayerState();
+    state.isPlaying = true;
+    state.shouldPlay = true;
+    let playerActions = getPlayerActions();
+    const {rerender} = render(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+    
+    state = getInitialPlayerState();
+    state.isPlaying = true;
+    state.shouldPlay = false;
+
+    let spyPause = jest.spyOn(media.players['0'], 'pause');
+    rerender(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+    expect(spyPause).toHaveBeenCalledTimes(1);
+  });
+
+  it('causes player to play when autoplay is set to true and state is active', () => {
+    let state = getInitialPlayerState();
+    let playerActions = getPlayerActions();
+    
+    const {rerender} = render(<MediaPlayer type={'video'} sources={getVideoSources()} playerState={state} playerActions={playerActions} />);
+    
+    let spyPlay = jest.spyOn(media.players['0'], 'playOrPlayOnLoad');
+        
+    rerender(
+      <MutedContext.Provider value={{muted: true}}>
+        <MediaPlayer type={'video'} sources={getVideoSources()} state={'active'} autoplay={true} playerState={state} playerActions={playerActions} />
+      </MutedContext.Provider>
+    );
+    setTimeout(()=>{
+      expect(spyPlay).toHaveBeenCalled();
+    }, 500);
```

Comment:
To use `setTimeout` in a test, the test needs to be async. Like this Jest does not know which test the `expect` belongs to. You can change it to `not.toHaveBeenCalled` and still see the test pass (without an assertion).

The core problem here, though, is the fact that we wrap the `MediaPlayer` in another component during the second render. This causes the `MediaPlayer` from the first render to be unmounted. The new `MediaPlayer` component then requests its own `player`. So `playOrPlayOnLoad` never gets called on the original player we spy on.

Reply:
Thanks alot for this